### PR TITLE
Sorted-index optimization for FastFieldTermSetQuery

### PR DESCRIPTION
**Pull request by RuchirRaj**
# Sorted-index optimization for FastFieldTermSetQuery

Optimizes `FastFieldTermSetQuery` execution on indexes sorted by the filter column. The goal is to make hash-join build-side InList pushdowns (`paradedb/paradedb#4895`) execute via index-aware traversal rather than a full fast-field scan.

## What changes

A new strategy framework in `term_set_query::term_set_strategy` decides per-segment how to evaluate a term set. Three strategies ship live; three more are stubs reserved for follow-ups.

**Live:**

- **Gallop** (sorted index + low K'/N): a streaming `TermSetGallopDocSet` walks the sorted column with a single forward cursor, lazily galloping to each term's contiguous range as `advance()`/`seek()` is called. Cost is `O(K · log(N/K))` per segment instead of `O(N)`. No upfront materialization of all K ranges — column reads happen only on cursor moves that cross a range boundary.
- **LinearScan** (fallback): existing `TermSetDocSet::advance` path, with a smart-seek override that jumps directly to a target DocId rather than walking forward one at a time.
- **AutomatonWeight** (K ≤ 1024): existing path, unchanged. Below this cardinality the strategy framework isn't engaged.

**Stubbed (route to LinearScan today):** `BitsetFromPostings`, `PostingListDirect`, `HashProbe`. The planner picks them based on K, D, and per-segment N inputs; dispatch unifies them with LinearScan until follow-ups land real implementations.

## Three implementation details worth flagging

**Lazy / streaming gallop, not eager union.** Earlier iterations of this PR materialized all K per-term DocId ranges upfront into a `RangeUnionDocSet`. The current design carries a single forward cursor and gallops on demand. `seek(target)` has a cheap in-range branch (no column read when `target` falls inside the cursor's current range) and a skip branch that uses `gallop_partition_point` over the term array to skip past intervening terms in `O(log K)` without ever reading their column ranges. This matters most for AND-intersection workloads where seeks dominate.

**Galloping search, not binary search.** Each per-term range lookup uses two-phase galloping (exponential probe from the cursor, then bounded binary search inside the bracketed interval) rather than plain bisection over the search window. Galloping wins because the early probes (`lo+1`, `lo+3`, `lo+7`) share a cache line with `lo`, while bisection misses cache on its first probe at the window midpoint. The constant factor compounds across thousands of per-term searches. Small windows (`hi - lo ≤ 32`) fall through to a batch-fetch into a stack `[u64; 32]` buffer plus linear scan, sidestepping per-element virtual dispatch on the fast tail. Helper-only bench: 1.65×–3.50× over plain bisection on the same workload.

**Smart-seek on `TermSetDocSet`.** The default `DocSet::seek` walks `advance()` until reaching the target. For AND-intersections where one side galloped down to sparse hits, this dilutes the selectivity benefit — the linear side re-traverses every doc to keep up. The override seeks directly to the target. Multi-column AND: 4–10× over default-seek depending on column density.

## Bench numbers

Single-column scan throughput, gallop vs LinearScan, on the LowFk corpus (D ≈ 100, the typical foreign-key shape):

| N | K | LinearScan | Gallop | Speedup |
|---|---|---|---|---|
| 10M | 10K | 267 MB/s | 3,350 MB/s | 13× |
| 10M | 100K | 232 MB/s | 606 MB/s | 2.6× |
| 50M | 10K | 267 MB/s | 10,800 MB/s | 40× |
| 50M | 100K | 242 MB/s | 1,910 MB/s | 7.9× |

PrimaryKey (D = 1) is faster still; HighFk (D ≈ 100K) sees smaller wins because the post-emission walk through the contiguous DocId ranges dominates per-term cost.

Two-column AND, smart-seek effect (N = 1M, K = 1500 per side):

| Fixture | Linear+Linear | Gallop+Linear | Speedup |
|---|---|---|---|
| Dense FK (D = 100) | 204 MB/s | 605 MB/s | 3.0× |
| Sparse PK (D = 1) | 157 MB/s | 1,750 MB/s | 11× |

These numbers are from synthetic LowFk-shaped corpora and were collected before the lazy/streaming redesign; the lazy path is expected to match or beat them on single-column scans (no upfront column reads) and improve further on multi-column AND (cheap in-range `seek`). Production DSG re-runs are tracked under `paradedb/paradedb#4895`.

Customer-typical workloads (N in 10M–50M, K in 1K–10K) are expected to land in the **40–90× scan throughput** improvement range. End-to-end wall-clock is Amdahl-capped by everything that isn't the scan.

## Defaults and observability

- `TermSetStrategyConfig` exposes per-strategy density thresholds; `gallop_max_density` defaults to `0.01` (1/100), bench-tuned on LowFk-shaped corpora. The threshold is checked per-segment against `K' / N`, where `K'` is the count of input terms surviving `[min, max]` pruning against the segment's column extents.
- `TermSetQuery::with_strategy_config` lets callers override defaults per-query.
- An optional `Arc` strategy sink lets callers (paradedb) read the per-segment dispatch decision back for `EXPLAIN ANALYZE` reporting. The sink is last-writer-wins across segments — the reported strategy is whichever segment finished selecting last, not an aggregate.
- `StrategyTag` is a `#[repr(u8)]` enum with `TryFrom` for the sink encoding.

## Tests

- Equivalence tests: gallop and LinearScan output identical DocId vectors across 7 corpus shapes (`tests/term_set_equivalence.rs`).
- Stress tests: `gallop_search_sorted` and `binary_search_sorted` agree on randomized inputs via `proptest`, including small-window batch-fetch paths and cursor-resumption invariants (`tests/gallop_search_stress.rs`).
- Property tests on `gallop_partition_point` and `gallop_search_sorted` in `src/query/range_query/sorted_internals.rs`.
- Unit tests on the strategy planner pin the dispatch shape across input combinations (`term_set_strategy::tests`).
- Direct unit tests on `TermSetGallopDocSet::advance`/`seek` covering cursor advancement, in-range seek, cross-range seek, and exhaustion.

## Follow-ups (out of scope here)

- Real implementations for `BitsetFromPostings`, `PostingListDirect`, `HashProbe` — extends the same speedup to unsorted indexes.
- Re-evaluating the `TERM_SET_FAST_FIELD_CARDINALITY_THRESHOLD = 1024` constant; current bench data suggests it bypasses gallop in regimes where gallop would win.
- D-aware threshold tuning: gallop's win shrinks with column duplication (D), but the current single-K/N threshold doesn't account for that.
- Per-segment strategy reporting: the current `AtomicU8` sink collapses across segments. A richer sink (e.g. per-segment vector or histogram) would let DSG benchmarking attribute wall-clock by dispatch fraction rather than inferring from baselines.

Refs paradedb/paradedb#4895.

### CODE REVIEW

File: src/query/range_query/sorted_internals.rs [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+    target: u64,
+    order: Order,
+    strict: bool,
+) -> u32 {
+    // Small-window fallback. With < 16 elements binary search does at most
+    // 4 comparisons; galloping pays one column read + branch per probe and
+    // can easily exceed that on tiny ranges.
+    if hi.saturating_sub(lo) < 16 {
+        return binary_search_sorted(column, lo, hi, target, order, strict);
+    }
```
Comment (stuhood) [inline src/query/range_query/sorted_internals.rs]: Is it even worth binary searching in this case, as opposed to linear scanning?

Note that `Column<u64>` does dynamic dispatch internally (to `ColumnValues`), so requesting a range of 16 contiguous values in one call with `column.first_vals` might be competitive with 4 random accesses.

You might also consider adding this case to the binary search: if the window being searched is smaller than N, fetch a batch of N contiguously.

Just watch out for allocation (use a thread local buffer or something).

File: src/query/range_query/sorted_internals.rs [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+
+    /// Randomized differential: drive both helpers with the same inputs
+    /// across many random configurations and assert agreement.
+    /// Complements the hand-written cases above — they pin specific
+    /// edge-cases; this one catches regressions in unanticipated regions
+    /// of the (lo, hi, target, order, strict) input space.
+    #[test]
+    fn randomized_differential_against_binary_search() {
+        use rand::prelude::*;
+        use rand::rngs::StdRng;
```
Comment (stuhood) [inline src/query/range_query/sorted_internals.rs]: Please use proptests (which allow for determinstic repro) rather than random inputs. See for example: https://github.com/paradedb/tantivy/blob/7aa40c6d0ad660233297e8a52bc23bb6506db64d/src/query/intersection.rs#L417-L439

File: src/query/term_set_query/term_set_query_fastfield.rs [hunk 3]
```diff
@@ -243,6 +302,36 @@ impl<T: Copy + Eq + std::hash::Hash + PartialOrd + std::fmt::Debug + Send + Sync
...
+    /// Jump the cursor directly to `target` and resume the per-doc match loop
+    /// from there. The trait default would call `advance()` in a loop and pay
+    /// one column read + one hash probe for every doc between the current
+    /// cursor and `target`; this override skips that work entirely.
+    ///
+    /// Important for multi-column AND-intersection (`BooleanQuery` of two
+    /// `TermSetQuery` filters): when one column gallops and the other lands
+    /// here, the intersection drives `seek(target)` calls toward the
+    /// gallop-emitted DocIds. Without this override, the dumb default seek
+    /// re-scans the gaps and the gallop speedup is largely diluted.
```
Comment (stuhood) [inline src/query/term_set_query/term_set_query_fastfield.rs]: Good catch!

Combining with ~any other `Query` type could have the same effect.
```suggestion
    /// Important for AND-intersection: when one column gallops and the other lands
    /// here, the intersection drives `seek(target)` calls toward the
    /// gallop-emitted DocIds. Without this override, the dumb default seek
    /// re-scans the gaps and the gallop speedup is largely diluted.
```

File: src/query/term_set_query/term_set_strategy.rs [hunk 4]
```diff
@@ -0,0 +1,655 @@
...
+/// the per-segment dispatch decision in `EXPLAIN ANALYZE` without depending
+/// on `TermSetStrategy` directly.
+pub mod strategy_tag {
+    pub const NONE: u8 = 0;
+    pub const GALLOP: u8 = 1;
+    pub const LINEAR: u8 = 2;
+    pub const BITSET: u8 = 3;
+    pub const POSTING: u8 = 4;
+    pub const HASH: u8 = 5;
+}
```
Comment (stuhood) [inline src/query/term_set_query/term_set_strategy.rs]: Please use a `repr(u8)` `enum` here: see https://github.com/paradedb/paradedb/pull/4939#discussion_r3170882616

File: src/query/term_set_query/term_set_query_fastfield.rs [hunk 5]
```diff
@@ -175,8 +198,44 @@ impl Weight for FastFieldTermSetWeight {
...
+
+                let strategy = select_strategy(
+                    reader,
+                    &column,
+                    PlannerInputs {
+                        field_name,
+                        candidate_size: None,
+                        avg_docs_per_term: None,
+                    },
+                    values,
```
Comment (stuhood) [inline src/query/term_set_query/term_set_query_fastfield.rs]: So, one thing to note here is that even the construction of this `HashSet` can be a significant portion of the total runtime: that's why it is important to include the construction time of the entire `TermQuery` in your benchmark (all the way up into `TermSet::new(...)`).

It's possible that it even makes sense to make the hashset construction itself lazy, so that if we choose a strategy based on sorting, we skip hashing.

File: tests/gallop_search_stress.rs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+                gallop_docs, linear_docs,
+                "differential mismatch: corpus(n={n}, seed={seed}, spread={spread}, \
+                 order={order:?}) target={target}",
+            );
+        }
+    }
+}
+
+#[test]
+fn gallop_path_matches_linear_path_on_multi_term_queries() {
```
Comment (stuhood) [inline tests/gallop_search_stress.rs]: Ditto proptests.

File: benches/term_set_queries.rs [hunk 7]
```diff
@@ -0,0 +1,677 @@
...
+    }
+    let reader = index
+        .reader_builder()
+        .reload_policy(ReloadPolicy::Manual)
+        .try_into()
+        .unwrap();
+    (reader.searcher(), field)
+}
+
+fn sample_terms(distinct: u64, k: usize, seed: u64) -> Vec<u64> {
```
Comment (stuhood) [inline benches/term_set_queries.rs:223]: Ditto proptests. Should not use `rand` at all... although I think that `proptest` has a way to generate random numbers: https://docs.rs/proptest/latest/proptest/prelude/trait.Rng.html

File: src/query/term_set_query/term_set_gallop.rs [hunk 8]
```diff
@@ -0,0 +1,492 @@
...
+/// `ContiguousDocSet`-wrapped scorers) because:
+///   - one allocation instead of K
+///   - the cursor + range index is a natural extension point for cursor- side optimizations (the
+///     `seek` walk is currently linear over the `ranges` vec; exponential probing inside `seek` is
+///     a future option if profiling shows it matters).
+pub(crate) struct RangeUnionDocSet {
+    /// Sorted ascending by `start`, non-overlapping. Empty ranges
+    /// (`start == end`) are filtered out at construction so iteration logic
+    /// can assume each entry contains at least one doc.
+    ranges: Vec<(DocId, DocId)>,
```
Comment (stuhood) [inline src/query/term_set_query/term_set_gallop.rs]: As mentioned offline, we should not precompute the entire set of ranges, as that will touch more buffers than needed when this Query is AND'd with something more selective: sorry, I think that the ticket had some bad guidance to that effect.

Instead, `advance` and `seek` should compute the beginning of the next range on the fly (possibly with some buffering or caching, but possibly not?).

If this feels like too big a change, we can punt it to a followup.

File: src/query/term_set_query/term_set_gallop.rs [hunk 9]
```diff
@@ -0,0 +1,492 @@
...
+    if non_null_start >= non_null_end {
+        return Box::new(EmptyScorer);
+    }
+
+    // Walk terms in the order that the column is sorted in. For ASC the input
+    // `sorted_terms` is already ascending; for DESC we iterate it in reverse
+    // so that the next matching term lives at higher DocIds than the previous
+    // one — `non_null_start` always increases.
+    let mut ranges: Vec<(DocId, DocId)> = Vec::with_capacity(sorted_terms.len());
+    let order_iter: Box<dyn Iterator<Item = &u64>> = match sort_order {
```
Comment (stuhood) [inline src/query/term_set_query/term_set_gallop.rs]: The `dyn` here will mean dynamic dispatch per call to `next`: instead, probably extract the loop as a generic function, and then call it with a concrete `I: Iterator<Item=&u64>`: https://stackoverflow.com/a/66597209

**Review by stuhood — APPROVED** [review]
Thanks! A few blocking comments, but #shipit so that we can iterate.

File: src/query/term_set_query/term_set_gallop.rs (hunk 8 above)
Comment (RuchirRaj) [reply src/query/term_set_query/term_set_gallop.rs]: I though of that too probably should do it, but in practice this is only used for the first column after which it's not sorted anyways, so might not really make a difference imo. 

I will just see if adding this makes the code harder to read or something, if not I will add it anyways

Comment (stuhood) [reply src/query/term_set_query/term_set_gallop.rs]: Any time the Query is ANDed with any other query this comes into play during calls to `seek`: this Query is still sorted, so it's still relevant. `seek` always moves forward.

Comment (RuchirRaj) [reply src/query/term_set_query/term_set_gallop.rs]: Yeah updated the logic

File: src/query/term_set_query/term_set_query_fastfield.rs (hunk 5 above)
Comment (RuchirRaj) [reply src/query/term_set_query/term_set_query_fastfield.rs]: Yeah now the hashset construction is now lazy

File: src/query/range_query/sorted_internals.rs [hunk 10]
```diff
@@ -0,0 +1,577 @@
...
+        column.first_vals(&docids[..window], &mut vals[..window]);
+        for i in 0..window {
+            // Safe: caller guarantees `[lo, hi)` is non-NULL.
+            let val = vals[i].expect("doc in non-NULL range has no value");
+            if !go_right(val, target, order, strict) {
+                return lo + i as u32;
+            }
+        }
+        return hi;
+    }
```
Comment (stuhood) [inline src/query/range_query/sorted_internals.rs:168]: Did this end up being noticeable faster? If not, feel free to remove.

File: src/query/term_set_query/term_set_gallop.rs [hunk 11]
```diff
@@ -0,0 +1,545 @@
...
+    next_term_idx: usize,
+    current_range_end: DocId,
+    current: DocId,
+}
+
+impl TermSetGallopDocSet {
+    pub(crate) fn new(
+        column: Column<u64>,
+        sort_order: Order,
+        sorted_terms: Vec<u64>,
```
Comment (stuhood) [inline src/query/term_set_query/term_set_gallop.rs]: Nit: if you mark an argument `mut`, then you can mutate it in place like a mutable local variable. Then you wouldn't need to define a new local for `let mut t` or `let terms` below.
```suggestion
        mut sorted_terms: Vec<u64>,
```

File: src/query/term_set_query/term_set_strategy.rs [hunk 12]
```diff
@@ -0,0 +1,685 @@
...
+
+    // Prune to [min, max].
+    let lo = column.min_value();
+    let hi = column.max_value();
+    let mut pruned: Vec<u64> = term_set
+        .iter()
+        .copied()
+        .filter(|v| *v >= lo && *v <= hi)
+        .collect();
+    let k_prime = pruned.len() as u32;
```
Comment (stuhood) [inline src/query/term_set_query/term_set_strategy.rs]: Could/should this be done in-place (mutating a `&mut Vec` argument), rather than into a collection that will be thrown away?

**Review by stuhood — APPROVED** [review]
Looks good: thank you!

File: src/query/range_query/sorted_internals.rs (hunk 10 above)
Comment (RuchirRaj) [reply src/query/range_query/sorted_internals.rs:168]: I did not benchmark it, actually, but it should be faster than the last implementation, and it's fine readability-wise also, so I don't think we need to remove it even if it's not significantly faster

File: src/query/term_set_query/term_set_gallop.rs (hunk 11 above)
Comment (RuchirRaj) [reply src/query/term_set_query/term_set_gallop.rs]: Done

File: src/query/term_set_query/term_set_strategy.rs (hunk 12 above)
Comment (RuchirRaj) [reply src/query/term_set_query/term_set_strategy.rs]: We won't be able to do it in place because the value is used across multiple segments, and each segment has a different min/max. 

But I updated it to only comput the count in general case and only allocate the vec when needed in the branch.
